src/model/android-versioning.js

```diff
@@ -0,0 +1,30 @@
+import * as semver from 'semver';
+
+export default class AndroidVersioning {
+  static determineVersionCode(version, inputVersionCode) {
+    if (!inputVersionCode) {
+      return AndroidVersioning.versionToVersionCode(version);
+    }
+    return inputVersionCode;
+  }
+
+  static versionToVersionCode(version) {
+    const parsedVersion = semver.parse(version);
+
+    if (!parsedVersion) {
+      return 1;
+    }
+
+    // The greatest value Google Plays allows is 2100000000.
+    // Allow for 4 patch digits, 2 minor digits and 3 major digits.
```

Comment:
To me 3-3-4 digits would make more sense than 3-2-4 digits.

Reply:
We only have room for 9 digits so unfortunately, that won't work. I prefer 3-2-4 over 3-3-3 because if people don't increment major/minor often, it's very convenient to have 4 digits. 

I use `${{ github.run_number }}` in my builds since it's a bit safer and it's not that important to have `versionCode` correspond to `versionName`. Preferences on this vary though.

Reply:
So if google play allows 2 100 000 000, that looks like 10 digits to me. Could you explain how you derive 9 digits?

My use case would be to have major always at 0 or 1 or 2. Perhaps if the project gets enormously successful i might get up to 20 in several years, but never 99, let alone need more digits than for minor.

> We only have room for 9 digits so unfortunately, that won't work. I prefer 3-2-4 over 3-3-3 because if people don't increment major/minor often, it's very convenient to have 4 digits.

Minor however I use when a new feature gets released. I will almost certainly reach 0.99.0 and 1.99.0. As major digit is reserved for semantically major releases I think 2 digits for minor level is not really safe either.

I would say that depending on the answer to the first question either 2-3-4 or 3-3-4 would be the best compromise. Let me know what you think.

Reply:
I'd really like to treat the most significant digit as reserved. When you get to 1000000000, that should be a strong warning sign to consider a different versioning scheme to avoid running out of numbers. 

With 3-2-4, I was thinking of apps like Firefox that increment the major monthly. We could also go 3-3-3 and aggressively throw errors when one of the numbers is over 999. In the future, there could also be an option for users to specify how the number is generated.

Reply:
I'll start us with `3-3-3`, it's backward compatible with `3-3-4` if you don't give any semantic meaning to the numbers.

>However we're not talking about enterprise software exposing a public API here but about builds for games or graphical programs that are generated by the Unity Editor. Versioning for games to me seems a lot more about having a unique incremental identifier for each build than about anything else.

I'm with you 100% on that. We don't actually use semver for our apps since there's really no consumer of that version and incrementing a number is a simpler mental model. The configuration we're using is

```
versioning: 'Custom'
version: 0.0.${{ github.run_number }}
androidVersionCode: ${{ github.run_number }}
```